include/revng/Pipeline/Rank.h

```diff
@@ -17,30 +18,96 @@ class Rank : public DynamicHierarchy<Rank> {
-  class RankDefinition : public Rank {
+namespace detail {
+
+template<typename... TypesToAppend, typename... InitialTupleTypes>
+inline constexpr auto appendImpl(std::tuple<InitialTupleTypes...> &&Tuple) {
+  return std::tuple<InitialTupleTypes..., TypesToAppend...>();
+}
+
+template<typename Tuple, typename... TypesToAppend>
+struct AppendToTupleHelper {
+  using type = decltype(appendImpl<TypesToAppend...>(std::declval<Tuple>()));
+};
+
+template<typename First, typename Second>
+struct CH {
+  template<std::size_t Index>
+  constexpr bool operator()() {
+    return std::is_convertible_v<std::tuple_element_t<Index, First>,
+                                 std::tuple_element_t<Index, Second>>;
+  }
+};
+template<typename Tuple>
+inline constexpr auto TS = std::tuple_size_v<Tuple>;
+
+// clang-format off
+template<typename LHS, typename RHS>
+concept TupleConvertibilityCheck = TS<LHS> == TS<RHS> &&
+                                   constexprAnd<TS<LHS>>(CH<LHS, RHS>{});
+// clang-format on
+
+} // namespace detail
+
+template<StringTemplateParameter Str, typename LocationType, typename ParentRnk>
+class RankDefinition : public Rank {
 public:
   static constexpr std::string_view RankName = Str;
   using Type = LocationType;
   using Parent = ParentRnk;
 
   using Definition = RankDefinition<Str, Type, Parent>;
+  using RankTag = Definition;
+
+public:
+  static_assert(std::is_same_v<Parent, typename Parent::Definition>);
+  static constexpr size_t Depth = Parent::Depth + 1;
+  using Tuple = typename detail::AppendToTupleHelper<typename Parent::Tuple,
+                                                     Type>::type;
 
 public:
   explicit RankDefinition(Parent &ParentObj) : Rank(RankName, ParentObj) {}
+
+  // clang-format off
+  template<typename ...Args>
+    requires detail::TupleConvertibilityCheck<std::tuple<Args...>, Tuple>
+  constexpr static Location<Definition> makeLocation(Args &&...Arguments) {
+    // clang-format on
+    return Location<Definition>(Arguments...);
+  }
+
+  constexpr static std::optional<Location<Definition>>
+  locationFromString(std::string_view String) {
+    return Location<Definition>::fromString(String);
+  }
 };
 
 // Root rank specialization
 template<StringTemplateParameter Str>
 class RankDefinition<Str, void, void> : public Rank {
 public:
   static constexpr std::string_view RankName = Str;
-  using Definition = RankDefinition<Str, void, void>;
   using Type = void;
   using Parent = void;
 
+  using Definition = RankDefinition<Str, Type, Parent>;
+  using RankTag = Definition;
+
+public:
+  static constexpr size_t Depth = 0;
+  using Tuple = std::tuple<>;
+
 public:
   explicit RankDefinition() : Rank(RankName) {}
+
+  constexpr static Location<Definition> makeLocation() {
+    return Location<Definition>{};
+  }
+
+  constexpr static std::optional<Location<Definition>>
+  locationFromString(std::string_view String) {
+    return Location<Definition>::fromString(String);
+  }
```

Comment:
I think ranks would benefit from being agnostic of `Location`.

Reply:
How do you suggest we construct locations then? Free helper-functions or something?

Reply:
Uhm, `Location<MyNiceRank>(...)`? Or maybe you can move this method as a `static` method in `Location`? Would that be possible?

Reply:
I guess that would work, I don't see why not. My main worry is that I don't really want a user having any control of the `Location`'s template argument. IMO it's much better to have
```
auto CoolLocation = revng::pipes::ranks::MyCoolRankDefinedInAHeader.makeLocation(...);
```
than
```
pipeline::Location CoolLocation(revng::pipes::ranks::MyCoolRankDefinedInAHeader, ...);
```
because then there's no ability explicitly pass anything as the `Location`'s template parameter.

Reply:
And by suggesting a free function, I meant something like
```cpp
auto CoolLocation = pipeline::makeLocation(revng::pipes::ranks::MyCoolRankDefinedInAHeader, ...);
```

Doesn't work with a static member though, because then there's again an opportunity to explicitly specify the template parameter.

Reply:
I'd prefer a `static` method if it's just an opportunity, but I'm fine with a free function too.
Let's avoid circular dependencies unless really needed.

Reply:
1. Well, it's an opportunity for a constructor (since those can deduce template parameter), for a static member it you'd have to specify those which is kind of painful.

2. There's no circular dependencies per se. Location relies on a `Rank-like` concept. And then the `RankDefinition` relies on the `Location`.

It's only now that I realized that I already have the system we've talked about in the `tag-vs-concept` thread in place - even if limited.

So, considering there's no circular dependencies, do you still prefer free-function-construction  over a static rank member?

Reply:
1. True, we need a free function.
2. Even if they are decoupled in the code, `Rank`-like is not really decoupled from `Rank`, it's pretty close to being a circular dependency. Conceptually, a `Location` needs a `Rank`, a `Rank` makes perfect sense in absence of `Location`s.

Reply:
Ok, yeah, you convinced me. Let's go with a free helper function.

Reply:
They are a set of free functions now.